core/src/main/java/com/linecorp/armeria/client/endpoint/SlowStartAwareEndpointSelectionStrategy.java

```diff
@@ -0,0 +1,135 @@
+     * </ul>
+     */
+    private final class SlowStartAwareWeightedRoundRobinSelector implements EndpointSelector {
+        private final EndpointGroup endpointGroup;
+        private final AtomicInteger sequence = new AtomicInteger();
+        private volatile EndpointsAndWeights endpointsAndWeights;
+        private final BiFunction<Endpoint, Duration, Integer> endpointWeighter;
+
+        private final Duration slowStartInterval = Duration.ofSeconds(10);
+        private final int maxWeight = 1000;
+        private volatile List<Endpoint> previousEndoints;
+        private volatile ScheduledFuture<?> updateEndpointWeightFuture = null;
+
+        SlowStartAwareWeightedRoundRobinSelector(EndpointGroup endpointGroup,
+                                                 BiFunction<Endpoint, Duration, Integer> endpointWeighter) {
+            this.endpointGroup = endpointGroup;
+            this.endpointWeighter = endpointWeighter;
+            final List<Endpoint> endpoints = endpointGroup.endpoints();
+            endpointsAndWeights = new EndpointsAndWeights(endpoints);
+            previousEndoints = endpoints;
+            endpointGroup.addListener(newEndpoints -> {
+                endpointsAndWeights = new EndpointsAndWeights(newEndpoints);
+                scheduleUpdateRequestWeight(newEndpoints, previousEndoints);
+                previousEndoints = endpoints;
+            });
+        }
+
+        @Override
+        public EndpointGroup group() {
+            return endpointGroup;
+        }
+
+        @Override
+        public EndpointSelectionStrategy strategy() {
+            return SlowStartAwareEndpointSelectionStrategy.this;
+        }
+
+        @Override
+        public Endpoint select(ClientRequestContext ctx) {
+            final int currentSequence = sequence.getAndIncrement();
+            return endpointsAndWeights.selectEndpoint(currentSequence);
+        }
+
+        private void scheduleUpdateRequestWeight(List<Endpoint> currentlyEndpoints,
+                                                 List<Endpoint> previousEndpoints) {
+            if (slowStartInterval.isZero()) {
+                return;
+            }
+            if (updateEndpointWeightFuture != null && !updateEndpointWeightFuture.isDone()) {
+                updateEndpointWeightFuture.cancel(true);
+            }
+            final Set<Endpoint> newEndpoints = Sets.difference(ImmutableSet.copyOf(previousEndpoints),
+                                                               ImmutableSet.copyOf(currentlyEndpoints));
+            final int step = Math.max((int) slowStartInterval.toMillis() / maxWeight, 1);
+
+            updateEndpointWeightFuture = clientFactory.eventLoopGroup()
+                                                      .schedule(() -> updateEndpointWeight(currentlyEndpoints,
+                                                                                           newEndpoints, step),
```

Comment:
I think each endpoint must be tracked individually because otherwise the weight of the previously newly added endpoints will jump to the target weight immediately, e.g.

1. A new endpoint A with weight 1000 is added.
2. Slow-start for the endpoint A starts.
3. Another new endpoint B with weight 1000 is added.
4. Slow-start for the endpoint B starts, but the weight of the endpoint A jumps straight up to 1000.